beacon-chain/p2p/peers/assigner.go

```diff
@@ -0,0 +1,72 @@
+package peers
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/cmd/beacon-chain/flags"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/sirupsen/logrus"
+)
+
+// handshakePollingInterval is a polling interval for checking the number of received handshakes.
+var handshakePollingInterval = 5 * time.Second
+
+func NewAssigner(ctx context.Context, s *Status, max int, finalized primitives.Epoch) *Assigner {
+	return &Assigner{
+		ctx:       ctx,
+		ps:        s,
+		max:       max,
+		finalized: finalized,
+	}
+}
+
+type Assigner struct {
+	sync.Mutex
+	ctx       context.Context
+	ps        *Status
+	max       int
+	finalized primitives.Epoch
+}
+
+var ErrInsufficientSuitable = errors.New("no suitable peers")
+
+func (a *Assigner) freshPeers() ([]peer.ID, error) {
+	required := params.BeaconConfig().MaxPeersToSync
+	if flags.Get().MinimumSyncPeers < required {
+		required = flags.Get().MinimumSyncPeers
+	}
+	_, peers := a.ps.BestFinalized(params.BeaconConfig().MaxPeersToSync, a.finalized)
```

Comment:
While it is a little annoying to make forkchoice's view of the finalized epoch visible to backfill in order to pass it to the Assign method, this is probably necessary in order to avoid picking a set of peers that agree on an older finalized epoch than our current view. So I'm pretty sure this needs to change to be given access to a method to get the latest finalized checkpoint.